geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/FileBasedMetadata.scala

```diff
@@ -92,151 +122,83 @@ class FileBasedMetadata(
       val envelope = EnvelopeConfig(partition.bounds.map(_.envelope).getOrElse(new Envelope()))
       PartitionConfig(partition.name, action, files, partition.count, envelope, System.currentTimeMillis())
     }
-    FileBasedMetadata.writePartitionConfig(fc, directory, config)
-    partitions.merge(partition.name, partition, FileBasedMetadata.remove)
+    val path = writePartition(config)
+    // if we have already loaded the partition, merge in the new value
+    if (metadata.getIfPresent(partition.name) != null) {
+      metadata.asMap.merge(partition.name, partition, removeMetadata)
+    }
+    Option(partitions.getIfPresent(BoxedUnit.UNIT)).foreach { files =>
+      files.merge(partition.name, PartitionFiles(config = Seq(config), parsed = Seq(path)), addFiles)
+    }
   }
 
-  override def compact(partition: Option[String], threads: Int): Unit = reload(partition, threads, write = true)
-
-  override def reload(): Unit = reload(None, 4, write = false)
-
-  override def close(): Unit = {}
-
-  private def reload(partition: Option[String], threads: Int, write: Boolean): Unit = {
+  override def compact(partition: Option[String], threads: Int): Unit = {
     require(threads > 0, "Threads must be a positive number")
-    val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(threads))
-    try {
-      // read the compaction baseline file
-      val compacted = FileBasedMetadata.readCompactedConfig(fc, directory)
-      // read any updates to the given partition(s)
-      // use a parallel collection, so that we get some threading on reading and deleting individual update files
-      val paths = FileBasedMetadata.listPartitionConfigs(ec, fc, directory, partition).par
-      paths.tasksupport = new ExecutionContextTaskSupport(ec)
-      val updates = paths.flatMap(FileBasedMetadata.readPartitionConfig(fc, _)).seq
-
-      partition match {
-        case None =>
-          // group by partition and merge updates by timestamp
-          val grouped = (compacted ++ updates).groupBy(_.name).values
-          val merged = grouped.flatMap(mergePartitionConfigs).filter(_.files.nonEmpty).toList
-
-          // update the in-memory map - update key by key, instead of clearing the map and repopulating it,
-          // so that callers don't ever see an empty map
-          val keys = new java.util.HashSet(partitions.keySet())
-          merged.foreach { m =>
-            keys.remove(m.name)
-            partitions.put(m.name, m.toMetadata)
-          }
-          keys.asScala.foreach(partitions.remove)
 
-          if (write) {
-            FileBasedMetadata.writeCompactedConfig(fc, directory, merged)
-            paths.foreach(fc.delete(_, false)) // note: already a parallel collection
-          }
+    // in normal usage, we never pass in a partition to this method
+    partition.foreach(p => logger.warn(s"Ignoring requested partition '$p' and compacting all metadata"))
```

Comment:
Admittedly, I'm skimming...  I do like the idea of being able to compact a range/partition/etc.  

Is that a sane addition or should it be a follow-on ticket?

Reply:
we never actually invoked it, so i didn't implement it this time. compacting a single partition means merging the partition-specific files into a compacted file that also contains other partitions, so it's a bit messy... and compacting region by region will end up rewriting the file over and over, so it's inefficient.
i think the reason the method has a partition at all is b/c it inherits from the Compactable trait that we created, otherwise i'd remove it. could maybe do something better around that trait, but eh